File: code/server/Project.cs
// Copyright (C) NOMAD Group <nomad-group.net>.
// For licensing information see LICENSE at the root of this distribution.

using protocol;
using System.Data.SQLite;
using System.IO;
using System.Threading.Tasks;

namespace noda
{ 
    // TEST:

    class Project
    {
        private SQLiteConnection connection;

        public string name { get; set; }
        public string md5 { get; set; }

        // new project
        public Project(string dbRoot, string name, string md5)
        {
            this.name = name;
            this.md5 = md5;

            // store the DB with md5 name to avoid collisions
            // TODO: what if there are several versions of the same _input_ file
            var cs = @"URI=file:" + Path.Combine(dbRoot, md5 + ".ndb");

            Logger.Trace("CS create: " + cs);
            connection = new SQLiteConnection(cs);
            connection.Open();
        }

        // existing project
        public Project(string fileName)
        {
            var cs = @"URI=file:" + fileName;
            connection = new SQLiteConnection(cs);
            connection.Open();

            // query the project name
            using (var cmd = new SQLiteCommand(
                "SELECT (value) FROM info WHERE name='idb_name'", connection))
            {
                using var reader = cmd.ExecuteReader();
                reader.Read();
                name = reader.GetString(0);
            }

            using (var cmd = new SQLiteCommand(
                "SELECT (value) FROM info WHERE name='idb_md5'", connection))
            {
                using var reader = cmd.ExecuteReader();
                reader.Read();
                md5 = reader.GetString(0);
            }
        }

        public void InitStorage()
        {
            using (var cmd = new SQLiteCommand(connection))
            {
                // create the project database layout
                cmd.CommandText =
                    @"CREATE TABLE info (
                            infoid INTEGER PRIMARY KEY AUTOINCREMENT, 
                            name TEXT NOT NULL, 
                            value TEXT NOT NULL
                      );
                      CREATE TABLE updates (
                            updateid INTEGER PRIMARY KEY AUTOINCREMENT, 
                            user_guid TEXT NOT NULL,
                            created TIMESTAMP DEFAULT CURRENT_TIMESTAMP, 
                            message_type BIGINT, 
                            data BLOB
                      );
                      CREATE TABLE users (
                            userid INTEGER PRIMARY KEY AUTOINCREMENT, 
                            user_guid TEXT NOT NULL, 
                            user_name TEXT NOT NULL
                      );

                      INSERT INTO info (name, value) VALUES('idb_name', @name);
                      INSERT INTO info (name, value) VALUES('idb_md5', @md5);";

                cmd.Parameters.AddWithValue("@name", name);
                cmd.Parameters.AddWithValue("@md5", md5);
                cmd.Prepare();

                cmd.ExecuteNonQuery();
            }
        }

        public void Update(string guid, Message msg)
        {
            using (var cmd = new SQLiteCommand(connection))
            {
                cmd.CommandText = @"INSERT INTO updates (user_guid, message_type, data) VALUES(@user_guid, @message_type, @data);";

                cmd.Parameters.AddWithValue("@user_guid", guid);
                cmd.Parameters.AddWithValue("@message_type", msg.MsgType);
                cmd.Parameters.AddWithValue("@data", msg.ByteBuffer.ToSizedArray());
                cmd.Prepare();

                cmd.ExecuteNonQuery();
            }
        }

        public void AddClient(string name, string guid)
        {
            using (var cmd = new SQLiteCommand(connection))
            {
                cmd.CommandText = @"INSERT INTO users (user_guid, user_name) VALUES(@user_guid, @user_name);";

                cmd.Parameters.AddWithValue("@user_guid", guid);
                cmd.Parameters.AddWithValue("@user_name", name);
                cmd.Prepare();

                cmd.ExecuteNonQuery();
            }
        }
    }
}

File: code/common/database/Project.cpp
// Copyright (C) NOMAD Group <nomad-group.net>.
// For licensing information see LICENSE at the root of this distribution.

#include <fmt/format.h>
#include <sqlite3.h>

#include "Project.h"

namespace database {

  Project::~Project()
  {
	if(_db)
	  sqlite3_close(_db);
  }

  bool Project::Create(const char *fileName)
  {
	if(sqlite3_open(fileName, &_db) != SQLITE_OK)
	  return false;

	// clang-format off
    static const char query[1024] =
	  R"(CREATE TABLE info (
            infoid INTEGER PRIMARY KEY AUTOINCREMENT, 
            name TEXT NOT NULL, 
            value TEXT NOT NULL
        );
        CREATE TABLE updates (
            updateid INTEGER PRIMARY KEY AUTOINCREMENT, 
            user_guid TEXT NOT NULL,
            created TIMESTAMP DEFAULT CURRENT_TIMESTAMP, 
            message_type BIGINT, 
            data BLOB
        );
        CREATE TABLE users (
            userid INTEGER PRIMARY KEY AUTOINCREMENT, 
            user_guid TEXT NOT NULL, 
            user_name TEXT NOT NULL
        );
        INSERT INTO info (name, value) VALUES('idb_name', ?);
        INSERT INTO info (name, value) VALUES('idb_md5', ?);)";
	// clang-format on

	sqlite3_stmt *stmt = nullptr;
	if(sqlite3_prepare_v2(_db, query, -1, &stmt, nullptr) != SQLITE_OK) {
	  sqlite3_close(_db);

	  // TODO: should really also delete the file...
	  return false;
	}

	sqlite3_bind_text(stmt, 1, name.c_str(), name.length(), nullptr);
	sqlite3_bind_text(stmt, 2, md5.c_str(), md5.length(), nullptr);
	sqlite3_step(stmt);
	sqlite3_finalize(stmt);
	return true;
  }

  void Project::AddMessage(const char *sourceId, const protocol::Message *message)
  {
	auto str = fmt::format(
	    "INSERT INTO updates (user_guid, message_type, data) VALUES({}, {}, {});",
	    sourceId, message->msg_type(), message->msg());
	sqlite3_exec(_db, str.c_str(), nullptr, nullptr, nullptr);
  }
} // namespace database